### ISSUE TITLE
Won't start on an offline PC

### ISSUE THREAD

**Issue by llvs** [MarkHopper24/barcodrod.io#25]
barcodrod.io installed properly on an offline Win11 machine from the github-downloaded install file but after clicking to start the app (no matter if immediately after installation or via Start menu) nothing happens. Only when the machine is connected to the internet (having provided the credentials in some other way), barcodrod.io opens up.
Does it need to fetch dependencies first?

The use case is to connect an offline laptop via webcam with a wifi network where the credentials come from a printed QR code. The built-in Win11 camera app can decode the QR code but needs a store app to handle it further and doesn't allow to copy the string (and therefore is totally useless).

I think this use case is important for barcodrod.io to shine as Win11 (with its camera app) doesn't help and your app could be the savior here.

**Comment by MarkHopper24** [comment]
Hi @llvs, thanks for the request. The GitHub hosted installer .msix is actually scraped from the MSFT Store. The main reason I went this route is because the Store package has the benefit of using the MSFT Store's code signing certificate, which I believe should be trusted on just about all Windows 10/11 machines. From a security perspective, I'm hesitant to publish an offline installer because it would be unsigned. Windows doesn't allow MSIX installations unless they're code-signed (without disabling security controls, using PowerShell, and enabling Developer Mode). 

Since I don't plan on purchasing one for this project since it's FOSS, it would be difficult to offer a .MSI or .EXE setup installer available securely since it would be unsigned.

To your issue- I suspect that since this is the Store version of the app, it's failing to launch for one of two reasons: 
Either it may need to contact Microsoft Store to acquire a license, or it's failing to install the Windows App Runtime which may need to be pulled from the Store. If you're able to either install the WinAppRuntime manually and/or launch the store while connected online once, I think it should work offline after that. FWIW, app usage can work completely offline. It's the install process that's a bit tricky.

I will investigate ways to package the .MSIX differently so it contains all relevant dependency packages for offline installer scenarios. In the meantime, the code is available so feel free to try those workarounds or build/modify locally for offline scenarios if you'd like.

**Comment by MarkHopper24** [comment]
Hi @llvs 

I believe I've found the root cause here, and it's because the .MSIX hosted in GitHub is the exact same as the Store version. Because it's the Microsoft-signed, store version, it requires an internet connection on installation for license acquisition through the Microsoft Store After first installation and launch, it should be able to be used fully offline. There's two options to solve this:

1. Release an unsigned version that needs to be sideloaded via PowerShell (not ideal- Edge/Defender will likely see it as unsafe). However, I could post checksums for users to self-verify.
2. I could acquire an offline, volume-licensed version of the Store buil through Winget, and provide that along with the .MSIX here. Only issue is these offline, volume-licensed versions are technically licensed against enterprises (you need to authenticate with Entra ID as an admin to download). I can acquire this just fine using my developer tenant, but I need to make sure that's not breaking any Terms & Conditions in the Microsoft Store.

TLDR; Two potential paths to fix, both with caveats.

**Comment by MarkHopper24** [comment]
@llvs May have a potential solution, can you please let me know if you would be willing to test? Thank you!

**Comment by FxRe** [comment]
I would love to see an offline installer for this, because MS Store is blocked in some environments. Even an unsigned version would be better than nothing. Apparently there are some free options for OSS projects like OSSign or SignPath?! Or maybe Sigstore is an option?

**Comment by MarkHopper24** [comment]
> I would love to see an offline installer for this, because MS Store is blocked in some environments. Even an unsigned version would be better than nothing. Apparently there are some free options for OSS projects like OSSign or SignPath?! Or maybe Sigstore is an option?

Yes- acknowledge the need and am exploring solutions. Appreciate those resources too, OSSign has applications paused but I've just submitted an application to SignPath, could be a path forward.

@llvs @FxRe 
I do have a question for you both. This makes sense in environments without the Windows store and for fully offline/isolated machines being used in organizations.

For those devices, organizations can use Winget to acquire an offline licensed version and install offline without the store. So for organizations, this is solved.

Are there folks who need an offline installer for personal use? Curious why?

**Comment by FxRe** [comment]
I’m trying to get this to run on a Windows Server 2019 Terminalserver that is online but does not have a working winget installation.

My use case is the following: we’re a small doctor’s office using rather mediocre barcode scanners that sometimes fail to read Data Matrix codes, especially from the screen when the codes appear in documents that we only have in digital form. I’m experimenting with barcodrod as a workaround in those cases, to make things a bit easier and to stop staff from printing pages of documents just so the hardware scanners can read them.

In our environment, the whole Microsoft Store and winget route is usually blocked completely because we do not use it, and our admins consider it a potential security risk. None of our other software requires it, so it is disabled via GPO and the necessary domains are not whitelisted in the firewall. They temporarily enabled and whitelisted it for me, which allowed me to install barcodrod on a local Windows 11 machine, but that does not help with the older Terminalserver, which I wanted to use for testing from home.

A standalone portable version that does not require installation, or alternatively a standalone installer, would work best for us. It would not have to be fully offline, as long as our admins know which domains need to be reachable for licensing and can whitelist them.

**Comment by MarkHopper24** [comment]
Hi @FxRe,

Appreciate the info.  To clarify, I’m not suggesting that winget needs to be available or usable on the target/client machine, like your Windows Server 2019 terminal server.
 
What I’m referring to is the admin-side offline distribution flow: an organization admin can use winget download from a managed/admin machine that has the required Store/Entra access to download the Store package, dependencies, and offline license file for distribution.
 
Microsoft documents this here:
 
  - winget download can download the installer, dependencies, and Microsoft Store offline license file: https://learn.microsoft.com/en-us/windows/package-manager/winget/download
  - Intune’s guidance covers this as the replacement path for offline Microsoft Store app distribution: https://techcommunity.microsoft.com/blog/intunecustomersuccess/downloading-microsoft-store-apps-using-windows-package-manager/4208020 (this all applies with or without Intune)
 
 Example flow would be something like:
 
  winget search "barcodrod.io" --source msstore
  winget download --id <package-id> --source msstore --download-directory <output-folder>
 
That should produce the app package, dependencies, and license artifacts that admins can then distribute through their normal software deployment process, without requiring each endpoint to use the Microsoft Store or run winget directly.
 
That said, I’m still exploring whether a separate portable or standalone installer makes sense, but for organizational offline deployment, the intended path is admin-created offline licensed distribution, not running winget on the client machine. I'll keep this issue updated as I explore alternatives, but sharing the details above since that would unblock your scenario.

**Comment by llvs** [comment]
> [@llvs](https://github.com/llvs) May have a potential solution, can you please let me know if you would be willing to test? Thank you!

Sorry for the late reply. Yes, I think I could test it, however the use case is not very complex and descibed in the OP: Make a notebook that has initially no internet connectivity able to join a wifi network with a very long and complicated password by showing a qr code through barcodrod.io that is installed through a usb key.

**Comment by FxRe** [comment]
Thank you for the detailed reply, @MarkHopper24!

What you're saying makes sense, I'll look into it. I know that we're not using Entra, though. If an Entra-ID is a requirement for downloading via winget, that might be another obstacle for us. I'll ask our sysadmins about it.

**Comment by MarkHopper24** [comment]
Hi @FxRe @llvs -- Good news, barcodrod.io has been accepted into SignPath's OSS Foundation and a fully offline, code-signed MSI release will be available soon. Appreciate the patience and willingness to help me understand your use cases!

**Comment by MarkHopper24** [comment]
@FxRe @llvs Happy to finally close this one out. v2.1 introduces (signed) offline installers for both x64 and Arm64. This package supports fully offline installation with no Store dependencies. 

https://github.com/MarkHopper24/barcodrod.io/releases/tag/v2.1
